From 7a0daf64b42de70e587073604e07050fbfce6b59 Mon Sep 17 00:00:00 2001
From: minaripenguin <minaripenguin@users.noreply.github.com>
Date: Fri, 16 Dec 2022 13:35:42 +0800
Subject: [PATCH] BootReceiver: Return early if trace_pipe doesn't exists

* instead of trigerring errnoexception, check if the file or directory exists first , otherwise perform a early return so we can skip the following exception below

12-16 13:05:46.592  1438  1438 E BootReceiver: Could not open /sys/kernel/tracing/instances/bootreceiver/trace_pipe
12-16 13:05:46.592  1438  1438 E BootReceiver: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at libcore.io.Linux.open(Native Method)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at libcore.io.BlockGuardOs.open(BlockGuardOs.java:274)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at android.system.Os.open(Os.java:494)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at com.android.server.BootReceiver.onReceive(BootReceiver.java:162)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:4336)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at android.app.ActivityThread.-$$Nest$mhandleReceiver(Unknown Source:0)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2154)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at android.os.Handler.dispatchMessage(Handler.java:106)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at android.os.Looper.loopOnce(Looper.java:201)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at android.os.Looper.loop(Looper.java:288)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at com.android.server.SystemServer.run(SystemServer.java:982)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at com.android.server.SystemServer.main(SystemServer.java:667)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at java.lang.reflect.Method.invoke(Native Method)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:553)
12-16 13:05:46.592  1438  1438 E BootReceiver: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:924)

Signed-off-by: minaripenguin <minaripenguin@users.noreply.github.com>
Signed-off-by: Dmitrii <bankersenator@gmail.com>
---
 services/core/java/com/android/server/BootReceiver.java | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/services/core/java/com/android/server/BootReceiver.java b/services/core/java/com/android/server/BootReceiver.java
index 572e9c2a5d82..065e398ad05d 100644
--- a/services/core/java/com/android/server/BootReceiver.java
+++ b/services/core/java/com/android/server/BootReceiver.java
@@ -139,6 +139,11 @@ public class BootReceiver extends BroadcastReceiver {
     private static final int MAX_ERROR_REPORTS = 8;
     private static int sSentReports = 0;
 
+    public boolean fileExists(String fileName) {
+       final File file = new File(fileName);
+        return file.exists();
+    }
+
     @Override
     public void onReceive(final Context context, Intent intent) {
         // Log boot events in the background to avoid blocking the main thread with I/O
@@ -161,6 +166,7 @@ public class BootReceiver extends BroadcastReceiver {
 
         FileDescriptor tracefd = null;
         try {
+            if (!fileExists(ERROR_REPORT_TRACE_PIPE)) return;
             tracefd = Os.open(ERROR_REPORT_TRACE_PIPE, O_RDONLY, 0600);
         } catch (ErrnoException e) {
             Slog.wtf(TAG, "Could not open " + ERROR_REPORT_TRACE_PIPE, e);
-- 
2.43.0

